### PR TITLE
Fix gateways discoveryAddress outside istio-system

### LINKED ISSUE #27265: Istio egressgateway outside istio-system does not support revision based install
**Bug description**

Installing istio with both egressgateway enabled and revision is set.

```
istioctl170 install  -f istio-egressgateway.yaml --set revision=istio170
```

istio-egressgateway.yaml:
```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  components:
    egressGateways:
      - name: "istio-egressgateway"
        enabled: true
        namespace: "istio-egress"
```
Found this log in egressgateway pod below, where `discoveryAddress: istiod.istio-system.svc:15012` seems incorrectly set to a non-revision based address.

```
$ kl -n istio-egress  istio-egressgateway-54c645697c-79jz2
Fri 11 Sep 2020 03:54:11 PM PDT: executing "kubectl logs -n istio-egress istio-egressgateway-54c645697c-79jz2"
2020-09-11T22:50:15.970698Z	info	FLAG: --concurrency="0"
2020-09-11T22:50:15.970765Z	info	FLAG: --disableInternalTelemetry="false"
2020-09-11T22:50:15.970775Z	info	FLAG: --domain="istio-egress.svc.cluster.local"
2020-09-11T22:50:15.970781Z	info	FLAG: --help="false"
2020-09-11T22:50:15.970786Z	info	FLAG: --id=""
2020-09-11T22:50:15.970790Z	info	FLAG: --ip=""
2020-09-11T22:50:15.970794Z	info	FLAG: --log_as_json="false"
2020-09-11T22:50:15.970799Z	info	FLAG: --log_caller=""
2020-09-11T22:50:15.970804Z	info	FLAG: --log_output_level="default:info"
2020-09-11T22:50:15.970820Z	info	FLAG: --log_rotate=""
2020-09-11T22:50:15.970825Z	info	FLAG: --log_rotate_max_age="30"
2020-09-11T22:50:15.970830Z	info	FLAG: --log_rotate_max_backups="1000"
2020-09-11T22:50:15.970835Z	info	FLAG: --log_rotate_max_size="104857600"
2020-09-11T22:50:15.970840Z	info	FLAG: --log_stacktrace_level="default:none"
2020-09-11T22:50:15.970852Z	info	FLAG: --log_target="[stdout]"
2020-09-11T22:50:15.970858Z	info	FLAG: --meshConfig="./etc/istio/config/mesh"
2020-09-11T22:50:15.970863Z	info	FLAG: --mixerIdentity=""
2020-09-11T22:50:15.970867Z	info	FLAG: --outlierLogPath=""
2020-09-11T22:50:15.970882Z	info	FLAG: --proxyComponentLogLevel="misc:error"
2020-09-11T22:50:15.970888Z	info	FLAG: --proxyLogLevel="warning"
2020-09-11T22:50:15.970894Z	info	FLAG: --serviceCluster="istio-egressgateway"
2020-09-11T22:50:15.970899Z	info	FLAG: --serviceregistry="Kubernetes"
2020-09-11T22:50:15.970903Z	info	FLAG: --stsPort="0"
2020-09-11T22:50:15.970908Z	info	FLAG: --templateFile=""
2020-09-11T22:50:15.970913Z	info	FLAG: --tokenManagerPlugin="GoogleTokenExchange"
2020-09-11T22:50:15.970919Z	info	FLAG: --trust-domain="cluster.local"
2020-09-11T22:50:15.970969Z	info	Version 1.7.0-2022348138e47498c4b54995b4cb5a1656817c4e-Clean
2020-09-11T22:50:15.971242Z	info	Obtained private IP [10.28.3.150]
2020-09-11T22:50:15.972648Z	info	Effective config: binaryPath: /usr/local/bin/envoy
concurrency: 0
configPath: ./etc/istio/proxy
controlPlaneAuthPolicy: MUTUAL_TLS
discoveryAddress: istiod.istio-system.svc:15012
drainDuration: 45s
envoyAccessLogService: {}
envoyMetricsService: {}
parentShutdownDuration: 60s
proxyAdminPort: 15000
serviceCluster: istio-egressgateway
statNameLength: 189
statusPort: 15020
terminationDrainDuration: 5s
tracing:
  zipkin:
    address: zipkin.istio-system:9411

```
This can also be reproduced with Isito168.

[ ] Docs
[x] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**

**Steps to reproduce the bug**

**Version (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version` if you used Helm)**
Istio 168
Istio 170

**How was Istio installed?**

**Environment where bug was observed (cloud vendor, OS, etc)**
GKE 1.16

### PR DESCRIPTION
**Pull request by elfinhe**
Please provide a description for what this PR is for.

Resolve: https://github.com/istio/istio/issues/27265

Tested by input IstioOperator:

```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  profile: empty
  components:
    ingressGateways:
    - name: default-ingressgateway
      enabled: true
      namespace: "gateways"
    egressGateways:
    - name: default-egressgateway
      enabled: true
      namespace: "gateways"
```
Install: `istioctldev install --set hub=gcr.io/istio-testing --set tag=1.9-dev --revision dev19 -f gateways-outside-istio-system.yaml`

Output of manifets:
https://gist.github.com/elfinhe/2dbb16471beb9f9e32efabdbd97087a3

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[x] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Review by howardjohn — CHANGES_REQUESTED** [review]
Shouldn't we put the full proxy config there? By doing this we give the false impression that this actually works, when in reality any proxy config field will not.

Note that any custom CA or XDS config, including external istiod and mesh ca, require proxy config customization, so it's a very common case

**Review by morvencao — COMMENTED** [review]
If the gws are installed into different namespace from the istiod, the mesh configmap can't be mounted into the gw pods:
https://github.com/istio/istio/blob/e60c24a399ab0e8293abfc74e9c7d7b7e7e4391e/manifests/charts/gateways/istio-ingress/templates/deployment.yaml#L331-L334
Do we need a copy of the mesh configmap when gws are installed into a different namespace?

**Comment by elfinhe** [comment]
> proxy config

Thanks John for the quick help. 

As morvencao also pointed out, do you agree if copying the ConfigMap istio-<REV> into the target namespace can be a more reasonable workaround?

btw I add this PR because I see more users reported the same issue happened to them in https://github.com/istio/istio/issues/27265.

**Comment by howardjohn** [comment]
@elfinhe I think we have a few options

* Add full proxy config into proxy config annotation. Downside: changes to mesh configmap directly have no effect, need to pass all the mesh config settings to gateway install
* Copy configmap to other namespaces. Downside: changes to mesh configmap directly have no effect, need to pass all the mesh config settings to gateway, depend on *how* we copy it it may be complex
* Inject the gateway. Downsides: new (incomplete) feature, more moving parts, untested. But it doesn't have any issues above

**Comment by elfinhe** [comment]
> @elfinhe I think we have a few options
> 
> * Add full proxy config into proxy config annotation. Downside: changes to mesh configmap directly have no effect, need to pass all the mesh config settings to gateway install
> * Copy configmap to other namespaces. Downside: changes to mesh configmap directly have no effect, need to pass all the mesh config settings to gateway, depend on _how_ we copy it it may be complex
> * Inject the gateway. Downsides: new (incomplete) feature, more moving parts, untested. But it doesn't have any issues above

@howardjohn for the short-term fix, I am thinking about the first option - proxy config annotation. 
Question - will proxy configs other than establishing Istiod/xds connections be pushed by istiod and override the ones in Istio agent during/after the bootstrapping? What's the source of truth?

I will update the PR with annotation copied from the config map. Just noticed that as for meshca workflow, proxyMetadata is already passed as envs for the istio-proxy in the gateway templates.

File: manifests/charts/gateways/istio-egress/templates/deployment.yaml [hunk 1]
```diff
@@ -49,6 +49,60 @@ spec:
         prometheus.io/path: "/stats/prometheus"
         {{- end }}
         sidecar.istio.io/inject: "false"
+        {{- if not (eq .Release.Namespace "istio-system") }}
+        proxy.istio.io/config: |-
+          {{- if .Values.global.meshID }}
```
Comment (howardjohn) [inline manifests/charts/gateways/istio-egress/templates/deployment.yaml:54]: I don't think we want to copy the entire thing here, as we will miss new fields. Also, I don't think this is correct as it doesn't handle the merging the we do in the real mesh config with `values.meshConfig`.

I think the way around this is to use templates in helm which I guess doesn't work with multiple charts... but I guess we could at least have a makefile component that syncs them (we do it for istio-remote today)

**Review by howardjohn — CHANGES_REQUESTED** [review]

**Pull request closed without merging**